apps/website/src/components/chart/hooks/use-chart-tooltip.ts

```diff
@@ -0,0 +1,117 @@
+import { useCallback } from 'react'
+
+import { localPoint } from '@visx/event'
+import { scaleTime } from '@visx/scale'
+import { useTooltip } from '@visx/tooltip'
+import { bisector, extent } from 'd3-array'
+
+import { getClosedIssues, getTotalIssues } from '../helpers/get-data'
+import { formatDate } from '../utils/format-time'
+import { calculatePercentage } from '../utils/get-percentage'
+
+import type { DayType } from '../chart'
+import type { EventType } from '@visx/event/lib/types'
+
+type Props = {
+  data: DayType[]
+  dates: Date[]
+  innerWidth: number
+  margin: {
+    top: number
+    right: number
+    bottom: number
+    left: number
+  }
+}
+
+/**
+ * An custom hook that handles the tooltip logic
+ * @param data - the data to be used in the chart
+ * @param margin - the margin of the chart
+ * @param dates - the dates of the chart
+ * @param innerWidth - the inner width of the chart
+
+ * @returns tooltipData - the data to be used in the tooltip
+ * @returns showTooltip - a function that shows the tooltip
+ * @returns hideTooltip - a function that hides the tooltip
+ 
+ **/
+
+const useChartTooltip = (props: Props) => {
+  const { data, margin, dates, innerWidth } = props
+  // tooltip parameters
+  const { tooltipData: tooltip, showTooltip, hideTooltip } = useTooltip()
+
+  const tooltipData = tooltip as DayType & {
```

Comment:
I am not completely sure how this works.

The `DayType` is defined as 
```ts
type DayType = {
  date: string
  open_issues: number
  closed_issues: number
}
```
but in the code you pass camel-cased keys.

Reply:
Also, `useTooltip` accepts generic type and allows you to pass initial state as an argument. 

Maybe this should be rewritten as
```ts
  const {
    tooltipData: tooltip,
    showTooltip,
    hideTooltip,
  } = useTooltip<DayType>({
    tooltipData: {
        completedIssuesPercentage: number
        openIssuesPercentage: number
        totalIssues: number
        openIssues: number
        closedIssues: number
        formattedDate: string
    },
  })
  ```
?

Reply:
This also correctly types the `showTooltip` function, which will be type-safe and requires you to pass all required fields.

Reply:
If i do this:

```
 // tooltip parameters
  const { tooltipData, showTooltip, hideTooltip } = useTooltip<TooltipData>({
    tooltipData: {
      date: '',
      completedIssuesPercentage: 0,
      openIssuesPercentage: 0,
      totalIssues: 0,
      openIssues: 0,
      closedIssues: 0,
      formattedDate: '',
    },
  })
  ```
 tooltipData will be `TooltipData | undefined`. The workaround is to assert the type of tooltipData:
 
   ```
   // Assert the type of tooltipData to TooltipData
  const tooltipDataAsserted = tooltipData as TooltipData
  ```